File: libraries/javaclient/src/main/java/com/ditrit/gandalf/java/event/ClientEvent.java
package com.ditrit.gandalf.java.event;

import com.ditrit.gandalf.java.grpc.*;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

public class ClientEvent {

    private String clientEventConnection;
    private String identity;
    private ManagedChannel channel;


    public String getClientEventConnection() {
        return clientEventConnection;
    }

    public void setClientEventConnection(String clientEventConnection) {
        this.clientEventConnection = clientEventConnection;
    }

    public String getIdentity() {
        return identity;
    }

    public void setIdentity(String identity) {
        this.identity = identity;
    }

    public ClientEvent(String senderEventConnection, String identity) {
        this.clientEventConnection = senderEventConnection;
        this.identity = identity;
        String[] connections = this.getClientEventConnection().split(":");
        this.channel = ManagedChannelBuilder.forAddress(connections[0], Integer.parseInt(connections[1]))
                .usePlaintext()
                .build();
    }

    public void SendEvent(String topic, String timeout, String uuid, String event, String payload) {
        ConnectorEventGrpc.ConnectorEventBlockingStub stub = ConnectorEventGrpc.newBlockingStub(channel);

        Empty empty = stub.sendEventMessage(EventMessage.newBuilder()
                .setTopic(topic)
                .setTimeout(timeout)
                .setUUID(uuid)
                .setEvent(event)
                .setPayload(payload)
                .build());
    }


    public EventMessage WaitEvent(String event , String topic, String idIterator) {
        ConnectorEventGrpc.ConnectorEventBlockingStub stub = ConnectorEventGrpc.newBlockingStub(channel);

        EventMessage eventMessage = stub.waitEventMessage(EventMessageWait.newBuilder()
                .setTopic(topic).setEvent(event).setIteratorId(idIterator)
                .build());
        return eventMessage;
    }

    public IteratorMessage CreateIteratorEvent() {
        ConnectorEventGrpc.ConnectorEventBlockingStub stub = ConnectorEventGrpc.newBlockingStub(channel);

        IteratorMessage iteratorMessage = stub.createIteratorEvent(Empty.newBuilder()
                .build());
        return iteratorMessage;
    }
}

File: libraries/goclient/event/ClientEvent.go
//Package event :
//File ClientEvent.go
package event

import (
	"context"
	"fmt"
	"time"

	pb "github.com/ditrit/gandalf/libraries/goclient/grpc"

	"github.com/google/uuid"
	"google.golang.org/grpc"
)

//ClientEvent :
type ClientEvent struct {
	ClientEventConnection string
	Identity              string
	client                pb.ConnectorEventClient
}

//NewClientEvent :
func NewClientEvent(identity, clientEventConnection string) (clientEvent *ClientEvent) {
	clientEvent = new(ClientEvent)
	clientEvent.Identity = identity
	clientEvent.ClientEventConnection = clientEventConnection

	conn, err := grpc.Dial(clientEvent.ClientEventConnection, grpc.WithInsecure())

	if err != nil {
		fmt.Println("clientEvent failed dial")
	}

	clientEvent.client = pb.NewConnectorEventClient(conn)
	fmt.Println("clientEvent connect : " + clientEvent.ClientEventConnection)

	return
}

//SendEvent :
func (ce ClientEvent) SendEvent(topic, event, referenceUUID, timeout, payload string) *pb.Empty {
	eventMessage := new(pb.EventMessage)
	eventMessage.Topic = topic
	eventMessage.Timeout = timeout
	eventMessage.UUID = uuid.New().String()
	eventMessage.Event = event
	eventMessage.Payload = payload
	eventMessage.ReferenceUUID = referenceUUID
	empty, _ := ce.client.SendEventMessage(context.Background(), eventMessage)

	return empty
}

//WaitEvent :
func (ce ClientEvent) WaitEvent(topic, event, referenceUUID, idIterator string) *pb.EventMessage {
	eventMessageWait := new(pb.EventMessageWait)
	eventMessageWait.WorkerSource = ce.Identity
	eventMessageWait.Topic = topic
	eventMessageWait.Event = event
	eventMessageWait.IteratorId = idIterator
	eventMessageWait.ReferenceUUID = referenceUUID
	eventMessage, _ := ce.client.WaitEventMessage(context.Background(), eventMessageWait)

	for eventMessage == nil {
		time.Sleep(time.Duration(1) * time.Second)
	}

	return eventMessage
}

//WaitEvent :
func (ce ClientEvent) WaitTopic(topic, referenceUUID, idIterator string) *pb.EventMessage {
	topicMessageWait := new(pb.TopicMessageWait)
	topicMessageWait.WorkerSource = ce.Identity
	topicMessageWait.Topic = topic
	topicMessageWait.IteratorId = idIterator
	topicMessageWait.ReferenceUUID = referenceUUID
	eventMessage, _ := ce.client.WaitTopicMessage(context.Background(), topicMessageWait)

	for eventMessage == nil {
		time.Sleep(time.Duration(1) * time.Second)
	}

	return eventMessage
}

//CreateIteratorEvent :
func (ce ClientEvent) CreateIteratorEvent() *pb.IteratorMessage {
	iteratorMessage, _ := ce.client.CreateIteratorEvent(context.Background(), new(pb.Empty))
	return iteratorMessage
}
